### ISSUE TITLE
Guides for creating Intel DPPY packages for conda-forge

### ISSUE THREAD

**Issue by PokhodenkoSA** [conda-forge/conda-forge.github.io#1449]
- [x] I read [the conda-forge documentation](https://conda-forge.org/docs/user/introduction.html#how-can-i-install-packages-from-conda-forge) and could not find the solution for my problem there.

Issue:

We want to start pushing out our packages `dpctl`, `numba-dppy`, `dpnp` to conda-forge to get make them widely available, and want some tips.

a) All of our packages will need the DPC++ compiler. I think we can handle it in our recipes by installing DPC++ (here is an example of what we currently do for dpctl on Github CI: https://github.com/IntelPython/dpctl/blob/48794f78206389f157ea2e86dbabb46fadbab6e8/.github/workflows/generate-coverage.yaml#L22) or explore using a prebuilt Docker image that has oneAPI preinstalled in our recipe.

Do you anticipate any problems with the usage of DPC++ and oneAPI in general in our recipes?

b) We doubt any of the CI systems that are available to feedstock will have Gen9 GPUs. We can of course build the packages without needing a GPU, but will not be able to test them on a GPU. An option can be to test them on a CPU SYCL device (using OpenCL or Level Zero CPU drivers) like we do in our GitHub Actions CI. 

Are you aware of any public CI service supports integrated GPUs? 

c) We do not support MacOS. Will that be a problem?

**Comment by beckermr** [comment]
Thanks for making this issue! Let's confer with @conda-forge/core

**Comment by beckermr** [comment]
> b) We doubt any of the CI systems that are available to feedstock will have Gen9 GPUs. We can of course build the packages without needing a GPU, but will not be able to test them on a GPU. An option can be to test them on a CPU SYCL device (using OpenCL or Level Zero CPU drivers) like we do in our GitHub Actions CI.
>
> Are you aware of any public CI service supports integrated GPUs?

No, we do not have access to such a system. We have been working through various means to enable this but we'll need outside sponsorship/support to make it happen.

> c) We do not support MacOS. Will that be a problem?

Not a problem at all.

**Comment by beckermr** [comment]
> We want to start pushing out our packages dpctl, numba-dppy, dpnp to conda-forge to get make them widely available, and want some tips.

Does `numba-dppy` mean that there will be two versions of numba on conda-forge? We should discuss this to ensure existing environments still work and that things are appropriately marked as conflicting to the solver, etc.

> a) All of our packages will need the DPC++ compiler. I think we can handle it in our recipes by installing DPC++ (here is an example of what we currently do for dpctl on Github CI: https://github.com/IntelPython/dpctl/blob/48794f78206389f157ea2e86dbabb46fadbab6e8/.github/workflows/generate-coverage.yaml#L22) or explore using a prebuilt Docker image that has oneAPI preinstalled in our recipe.

We have been coordinating with another group at intel on the compilers. I sent this issue to them but cannot seem to find their github handles right now. We should get their opinions here too.

**Comment by diptorupd** [comment]
> No, we do not have access to such a system. We have been working through various means to enable this but we'll need outside sponsorship/support to make it happen.

One option is to use [Intel DevCloud](https://software.intel.com/content/www/us/en/develop/tools/devcloud.html).  Intel DevCloud has Intel GPU nodes and is free for use via SSH. We can potentially (at least for Github CI) connect to DevCloud and use a GPU node there. I have yet to try it out, so not sure of latency and node availability. I will try it and post an update.

**Comment by diptorupd** [comment]
> Does `numba-dppy` mean that there will be two versions of numba on conda-forge? We should discuss this to ensure existing environments still work and that things are appropriately marked as conflicting to the solver, etc.

No, `numba-dppy` is a standalone extension/plug-in to Numba. `Numba-dppy` is only needed to get access to the code-generator for SYCL devices.

**Comment by diptorupd** [comment]
> > a) All of our packages will need the DPC++ compiler. I think we can handle it in our recipes by installing DPC++ (here is an example of what we currently do for dpctl on Github CI: https://github.com/IntelPython/dpctl/blob/48794f78206389f157ea2e86dbabb46fadbab6e8/.github/workflows/generate-coverage.yaml#L22) or explore using a prebuilt Docker image that has oneAPI preinstalled in our recipe.
> 
> We have been coordinating with another group at intel on the compilers. I sent this issue to them but cannot seem to find their github handles right now. We should get their opinions here too.

+ Nikolay Petrov (@napetrov)

**Comment by AndresGuzman-Ballen** [comment]
We will be publishing the DPC++ compiler conda packages for 2021.3 release in a few weeks. There are 2021.2 versions that are in a test channel, but in any case it should work just fine by adding the following to the meta.yaml:
`requirements:
   build:
      - {{  compiler('dpcpp')  }}
`

**Comment by beckermr** [comment]
We recently removed defaults from our default channels list for builds. It should be possible to add it back, but I am sure how to do that on staged recipes.

**Comment by PokhodenkoSA** [comment]
Thank you for your feedback.

**Comment by isuruf** [comment]
We should first package the compilers and runtimes in conda-forge before creating downstream packages.
